libraries/AC_PID/AC_PID_2D.cpp

```diff
@@ -121,24 +121,49 @@ Vector2f AC_PID_2D::update_all(const Vector2f &target, const Vector2f &measureme
     return _error * _kp + _integrator + _derivative * _kd + _target * _kff;
 }
 
-Vector2f AC_PID_2D::update_all(const Vector3f &target, const Vector3f &measurement, bool limit)
+Vector2f AC_PID_2D::update_all(const Vector3f &target, const Vector3f &measurement, const Vector3f &limit)
 {
-    return update_all(Vector2f{target.x, target.y}, Vector2f{measurement.x, measurement.y}, limit);
+    return update_all(Vector2f{target.x, target.y}, Vector2f{measurement.x, measurement.y}, Vector2f{limit.x, limit.y});
 }
 
 //  update_i - update the integral
-//  If the limit flag is set the integral is only allowed to shrink
-void AC_PID_2D::update_i(bool limit)
+//  If the limit is set the integral is only allowed to reduce in the direction of the limit
+void AC_PID_2D::update_i( const Vector2f &limit)
 {
-    float integrator_length_orig = _kimax;
-    if (limit) {
-        integrator_length_orig = MIN(integrator_length_orig, _integrator.length());
-    }
-    _integrator += (_error * _ki) * _dt;
-    const float integrator_length_new = _integrator.length();
-    if (integrator_length_new > integrator_length_orig) {
-        _integrator *= (integrator_length_orig / integrator_length_new);
+    Vector2f limit_direction = limit;
+    Vector2f delta_integrator = (_error * _ki) * _dt;
+    if (!is_zero(limit_direction.length_squared())) {
+        // zero delta_vel if it will increase the velocity error
+        limit_direction.normalize();
+        if (is_positive(delta_integrator * limit)) {
+            delta_integrator.zero();
+        }
     }
+
+    _integrator += delta_integrator;
+    _integrator.limit_length(_kimax);
+}
+
+Vector2f AC_PID_2D::get_p() const
+{
+    return _error * _kp;
+}
+
+Vector2f AC_PID_2D::get_i() const
```

Comment:
return _derivative * _kd;

I think we can only do I right?

Reply:
Compile errors meant I could not do this.

Reply:
OK, so I omitted `const` - it does work like this:

```
-    Vector2f get_i() const;
+    const Vector2f& get_i() const;
```

(because the function is const you can't return a reference to the object without also marking that as `const`.

And yes, can only do I - I misread.